Fixed recompute grad issue - no memory savings

pidajay opened the pull request:
This PR addresses an issue where no memory saving was observed with gradient checkpointing in eager mode.
Results can be found here - https://github.com/pidajay/tf2_gradient_checkpointing/blob/master/Results_tf_recompute_grad_bug_fix.txt
Summary: Peak memory for a sample CNN model before fix is 3.3 GB and after fix is 704 MB.
This is the script used to generate results - https://github.com/pidajay/tf2_gradient_checkpointing/blob/master/run_recompute_grad_with_profile.py.

Commit:
Fixed eager mode gradient checkpointing by eliminating unecessary persistence of intermediate activations in memory

pidajay commented:
There seems to be a lack of documentation on how to use this feature. I have a tutorial in works over here - https://github.com/pidajay/tf2_gradient_checkpointing/blob/master/tf_recompute_grad_tutorial.ipynb. I would be happy to add it to tensorflow docs if someone can point me to the right location.

pidajay commented:
The use of this feature depends on the user having to manually partition the model. This can be an unpleasant user experience in many cases. For example someone might just be interested in performing transfer learning by leveraging a very large pre-trained model. I have created a github issue with preliminary solution over here - https://github.com/tensorflow/tensorflow/issues/38766

alextp commented on tensorflow/python/ops/custom_gradient.py:
```diff
@@ -406,14 +406,17 @@ def internal_grad_fn(unused_op, *result_grads):  # pylint: disable=unused-variab
-    result, grad_fn = f(*args, **kwargs)
+
+  trainable_vars = []
+  if 'trainable_variables' in kwargs:
```
This is an implicit API change where you now ask users to explicitly pass trainable_variables to the decorated function.

Can you also edit the documentation? Can you add tests that show behavior which is broken without this explicit hatchet? Can you clarify what's the plan for code that creates variables when it's called?

pidajay commented:
@alextp  Looking for guidance on a strategy to test this. The benefits of this feature can only be seen on large models that generate huge number of intermediate activations. I can think of two ways
1. Train one step without grad checkpointing on a GPU that causes OOM. Then successfully train the same model with grad checkpointing on same GPU without OOM. But this approach assumes specific hardware config for running tensorflow tests which may not be desirable.
2. Train one step on CPU with and without grad checkpointing and profile the memory. This is what I have shown [here](https://github.com/pidajay/tf2_gradient_checkpointing/blob/master/Results_tf_recompute_grad_bug_fix.txt). But again this involves training a large model in a tensorflow test on a CPU which is not desirable.
I am not sure if there exists an easy way to profile GPU memory. Then step 2 on a GPU might be plausible. Any thoughts?

alextp commented:
You're proposing a change to TF's public API and for us to review it you should be able to write documentation specifying how this change needs to be used and to write a unit test that exercises this change.

I will say though that I do not like the API change, and I also don't understand why replacing the variable watcher with an explicit list of variables solves the problem. Can't we fix the problem by fixing the variable watcher without any API changes?

pidajay commented:
@alextp . Thanks for your comments and appreciate your patience, especially since this is my first PR to TF core :).
I understand API change is a hassle. The hatchet was definitely required in TF2.1 where there was no variable_watcher and tape.watch was storing all the intermediate outputs while tape.stop_recording would not give me the trainable variables. It is quite possible this may not be the case anymore with variable_watcher. I will try to explore ways to avoid this API change and build a more solid case. Thanks!

alextp commented:
Thanks! Please keep me updated on your progress, and post here if you get
stuck.

On Mon, May 4, 2020 at 2:39 PM pidajay <notifications@github.com> wrote:

> @alextp <https://github.com/alextp> . Thanks for your comments and
> appreciate your patience, especially since this is my first PR to TF core
> :).
> I understand API change is a hassle. The hatchet was definitely required
> in TF2.1 where there was no variable_watcher and tape.watch was storing all
> the intermediate outputs while tape.stop_recording would not give me the
> trainable variables. It is quite possible this may not be the case anymore
> with variable_watcher. I will try to explore ways to avoid this API change
> and build a more solid case. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/39042#issuecomment-623722086>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRNGMDRWGIKABU5GF6TRP4YZLANCNFSM4MUEMYXQ>
> .
>


-- 
 - Alex

Commit:
Removed API change and added tests.

pidajay commented:
@alextp Turns out API change is not necessary. The introduction of variable watcher post tf 2.1 voids the need for this. I have updated the PR and added tests to show the efficacy of the fix. Additional memory profile results can be found [here](https://github.com/pidajay/tf2_gradient_checkpointing/blob/master/Results_tf_recompute_grad_bug_fix.txt). If the code changes look ok I can look into fixing the doc as the current doc is quite sparse with no explanation on how to use this feature. Thanks!

alextp commented on tensorflow/python/keras/integration_test/gradient_checkpoint_test.py:
```diff
@@ -0,0 +1,160 @@
+    with self.assertRaises(Exception) as context:
+      _train_no_recompute(1)
+    self.assertTrue(
+        context.exception.__class__.__name__ == 'ResourceExhaustedError')
```
I am scared of tests that check for actual OOMs as they are sensitive to things like what specific device it's run on, etc.

I see two options here:
 1. use a virtual GPU device with fixed limited memory
 2. (my preferred alternative) use side effects to detect whether things are recomputed. For example if the forward pass does variable.assign_add(1) we can count how many times it ran

pidajay replied:
I am using a virtual GPU if you look at line 76 inside the _limit_gpu_memory() function. Maybe I should bazel tag these tests to run only on a virtual GPU? I actually like the OOM test since it demonstrates and confirms that the grad checkpointing is functional. Otherwise it might remain silent and broken like it has been in TF 2.x all along. But happy to roll it back and add your preferred option 2 if you feel this is risky.

alextp replied:
Ah I missed it. This is good then. Thanks!

alextp reviewed (dismissed):
Thanks for the fix!

rthadur commented:
@pidajay can you please check build failures.

pidajay commented:
@rthadur gradient_checkpoint_test.py fails because it needs to run only on GPU. I guess I need to tag it as a cuda_py_test. I will investigate the other failing test (forward_prop_test) tonight. Thanks!

pidajay commented:
@alextp My minor modification to recompute_grad seems to have impacted testCustomGradientRecomputeGrad(). This test tries to use recompute_grad to compute gradients in forward mode differentiation. I am not an expert but intuitively this does not make sense to me. The point of recompute_grad is to save memory while using reverse mode auto diff. My options
1. Eliminate the test if it serves no purpose. 
2. If the test does serve a purpose then I got a problem on my hands. Gradient checkpointing saves memory by not saving intermediate activations. So it is fundamentally incompatible with forward mode diff. Back to the drawing board?

Thanks!

alextp commented:
@allenlavoie the comment by @pidajay makes me think we should delete that test. WDYT?

allenlavoie commented:
To be clear that test isn't about saving memory, just that forwardprop doesn't crash and gives the right answer with recompute_grad. The program in that test is not even crashing, it's just giving the wrong answer (0) silently. So just deleting the test is definitely not the right thing to do.

It's catching a correctness issue here that's not related to forwardprop infrastructure (i.e. a user could reproduce it with some GradientTapes). With this change we're basically inserting some stop_gradients that define gradients with respect to args/kwargs of the custom gradient function to be zero when they aren't.

We don't have to support transposing recompute_grad-decorated things on a tape necessarily, so defining a new inner nested custom_gradient that throws an error here is an option (in which case please just modify the test to look for the error instead of a correct gradient). Or we could define an inner nested custom_gradient that gives the right answer (and maybe throw a warning that it's probably not saving memory?). `tf.custom_gradient`'s docstring mentions how to define gradients for forwardprop: https://www.tensorflow.org/api_docs/python/tf/custom_gradient

pidajay commented:
@allenlavoie Thanks for the detailed response. I like both your suggestions :). However for me to execute any of those recompute_grad will need to know if it is in forward or reverse mode. I can think of doing it two ways, explicitly through an optional arg or implicitly through **kwargs. Either ways wouldn't this constitute an API change and isn't that a problem?

allenlavoie commented:
You can do all of this with just more custom_gradient decorators:

```
import tensorflow as tf

@tf.custom_gradient
def original_f(x):
  y = x * 2.

  @tf.custom_gradient
  def f_grad(dy):

    @tf.custom_gradient
    def f_grad_impl(x_dupe):

      def second_order(ddy):
        raise ValueError("You tried to take a second-order gradient of original_f")

      return dy * 2, second_order

    def transpose(ddy):
      raise ValueError("You tried to transpose original_f!")

    return f_grad_impl(x), transpose

  return y, f_grad

with tf.GradientTape(persistent=True) as tape:
  x = tf.constant(1.)
  tape.watch(x)
  y = original_f(x)
  grad = tape.gradient(y, x)

  try:
    grad_grad = tape.gradient(grad, x)
  except ValueError as e:
    assert "second-order" in str(e)

  tangent = tf.constant(1.)
  tape.watch(tangent)
  grad = tape.gradient(y, x, tangent)

  try:
    transpose = tape.gradient(grad, tangent)
  except ValueError as e:
    assert "transpose" in str(e)
```

The "transpose" path is what is being triggered by the forwardprop test. But it's also checking second-order gradients, so you may have to do something similar there (do we not have other tests for second-order gradients of recompute_grad?).

allenlavoie commented:
Actually second-order gradients look like they're fine. So maybe ignore that part.

gbaned commented:
@pidajay Can you please check @allenlavoie's comments and keep us posted. Thanks!

pidajay commented:
@gbaned Planning to update the PR this afternoon. Thanks!

Commit:
Modified recompute_grad to handle fwd mode diff

pidajay commented:
@allenlavoie Thanks for the explanation and the code snippet. That clarified things for me. I have tried to incorporate your suggestions. Please let me know if they look ok.

allenlavoie commented on tensorflow/python/ops/custom_gradient.py:
```diff
@@ -482,27 +482,42 @@ def recompute_grad(f):
+      def transpose(*t_args, **t_kwargs):
+        """Gradient function calculation for forward mode autodiff."""
+        # Just throw an error since gradients / activations are not stored on tape for recompute.
+        raise ValueError(
```
NotImplementedError? We can go back to implementing this in the future. And if someone is relying on it we'll have to.

pidajay replied:
Agreed. Will fix it.

pidajay commented on tensorflow/python/eager/forwardprop_test.py:
```diff
@@ -357,7 +364,10 @@ def testCustomGradientRecomputeGrad(self):
     def f(x):
```
This test seems to fail for me locally because of the decorator - @test_util.assert_no_new_pyobjects_executing_eagerly. I am currently investigating it.

allenlavoie replied:
I vaguely remember having problems with assertRaises and some of the memory decorators. Which objects does it say are leaked?

pidajay replied:
AssertionError: The following objects were newly created: Counter({'cell': 30, 'function': 21, 'tuple': 18, 'list': 9})

allenlavoie replied:
Yeah I wouldn't worry about it for this change. You can convert the test to checking for the error and then remove the memory decorator.

It's possible there's a failure to free references when returning an exception through the forward accumulator or through the tape. My first thought was that assertRaisesRegexp causes an issue, but we do have tests that use it with the memory checking decorator. Can you leave a `TODO(allenl): investigate why assert_no_new_pyobjects_executing_eagerly fails around this test`?

pidajay replied:
Sounds good. Thanks for looking into it. It was troubling me quite a bit :). Will put the todo in my upcoming commit.

Commit:
Addressed PR comments

allenlavoie commented on tensorflow/python/ops/custom_gradient.py:
```diff
@@ -482,26 +482,42 @@ def recompute_grad(f):
-        t.watch(id_args)
+      """Nested custom gradient function for computing grads in reverse and forward mode autodiff."""
+
+      def grad_eval():
```
Any reason to have this inner nested function? I think you can take it out and save some indentation

pidajay replied:
No reason. Removed it.

Commit:
Addressed PR comments

pidajay commented:
How do I specify a test to be GPU only? I thought the label cuda_py_test will automatically do it for you. But it looks the the internal CI system tries to run it on CPU too.

allenlavoie commented:
You can skip the test on CPU, like this: https://github.com/tensorflow/tensorflow/blob/f38355dab31bb466e9fdc900089dcd4abba536d6/tensorflow/python/eager/tensor_test.py#L80-L82

allenlavoie commented:
Actually can you add a test for higher-order backward gradients too? Like I say based on what was failing in _test_gradients in the forwardprop tests they're probably fine, but as far as I can see we're not testing them. You can copy _test_gradients and omit the forward-mode bits and we can use that in gradients_test.py if you want.

pidajay commented:
Sorry I think I missed the last part. If I understood it right, we are not explicitly testing the higher order backward gradients since I moved the assertRaisesRegexp around the whole of _test_gradients. So you are suggesting I create another _test_gradients to just test the backward gradients and possibly move this method into /tensorflow/python/ops/gradients_test.py?

allenlavoie commented:
Oh, right, that's why you had the argument in the forwardprop test. Sorry for the ping-ponging. Yeah I think that's probably cleaner than having a backward-only test in the forwardprop test suite. I certainly wouldn't think to look there. If you remove the forwardprop bits _test_gradients should be trivial to copy I think (just _grad and _test_gradients?).

If you want you can move the transpose error message test next to the other recompute_grad tests too. It doesn't need the forwardprop _test_gradients implementation to reproduce, you can either use a GradientTape like in my example or just a manual ForwardAccumulator.

pidajay commented:
Ok that sounds good to me. Will work on it and update the PR possibly tomorrow morning. Thanks for your patience and appreciate all the suggestions.

Commit:
Reorganized tests for recompute grad

pidajay commented:
@allenlavoie I reorganized the tests as you suggested. Sorry for the delay. I am having trouble executing the tests on my setup. Some of these tests fail locally in graph mode even prior to any of my changes. I noticed some of the gradients tests have been decorated with @test_util.deprecated_graph_mode_only. Also the bazel label for the entire gradients test target is 'no_oss'. Not sure if these tests meant to be executed on the github CI?

allenlavoie commented:
Oh wow, disabled 2 years ago because of flakiness on GPUs. Yeah, with no_oss they won't run on the github CI. The deprecated_graph_mode_only decorators mean we don't run them with 2.x behavior, probably because they assume Graph+Sessions. I wouldn't add it to new tests, but we do still run them in some places.

You can start a new test suite (custom_gradient_test.py ?) if you want to work around the no_oss tag for now. But if you add it to that file we do have CI builds that run those tests still. I tried and they do run internally; which errors are you seeing?

pidajay commented:
These are the errors I see which seem to predate my changes. But if they do run successfully on your internal system, then I will re-verify on my end.

ERROR: testFnRecomputeWithScopeGradientTape (__main__.VariablesGradientTest)
ValueError: The custom_gradient decorator currently supports keywords arguments only when eager execution is enabled.

ERROR: testFnRecomputeWithScopeGradients (__main__.VariablesGradientTest)
ValueError: The custom_gradient decorator currently supports keywords arguments only when eager execution is enabled.

ERROR: testFnRecompute (__main__.VariablesGradientTest)
tensorflow.python.framework.errors_impl.FailedPreconditionError: Error while reading resource variable test/test_var from Container: localhost. This could mean that the variable was uninitialized. Not found: Container localhost does not exist. (Could not find resource: localhost/test/test_var)
	 [[{{node test/ReadVariableOp}}]]

pidajay commented:
My bad. Looks like the some of the failed tests are indeed capturing a problem. In my quest to make things work in eager mode, I never paid attention to graph mode. Seems like the design to use nested custom gradients is not sitting happy with the graph_mode_decorator. Specifically graph mode decorator is not happy with kwargs while we do have to make use of kwargs in the nested calls.

allenlavoie commented:
Ugh, that's annoying. Is it just the `variables` argument? I think if you take that explicitly it'll be OK

Otherwise you can work around it with `nest.flatten` and `nest.pack_sequence_as` (pass arguments through custom_gradient as a flat list, re-structure them inside).

pidajay commented:
It is not just that. I think the root cause has something to do with the use of resource variables. Somehow it does not seem happy with the resource variables that appear in the nested call.

This is the mini-test I run
```
# this passes
def testFnRecompute(self):
	test_input = constant(np.zeros((1, 2), dtype=np.float32))
	grads_re, grads = self._TestFnVariablesGradient(test_input, TestFn, test_input)

# this will fail
def testFnRecompute(self):
	test_input = constant(np.zeros((1, 2), dtype=np.float32))	
	with variable_scope.variable_scope("test", use_resource=True):
             test_var = variable_scope.get_variable(name="test_var",
             shape=2, trainable=True)
        grads_re, grads = self._TestFnVariablesGradient(test_input, TestFn, test_var)
```
It fails with this error message.
tensorflow.python.framework.errors_impl.FailedPreconditionError: Error while reading resource variable test/test_var from Container: localhost. This could mean that the variable was uninitialized. Not found: Container localhost does not exist. (Could not find resource: localhost/test/test_var)
	 [[{{node test/ReadVariableOp}}]] 
This seems to be the line in the graph mode decorator that triggers it. 
``` args = [ops.convert_to_tensor(x)for x in args] ```

allenlavoie commented:
Oh, ha, so the test just isn't initializing its variable, and previously that was fine because the value was never used? I think it's fine to insert a self.evaluate(test_var.assign(tf.zeros([10]))) to initialize it if so.

Commit:
Fixed tests

pidajay commented:
Thanks @allenlavoie. That fixed it. Looks like working with eager mode has spoilt me :). I have updated the PR. I had to re-wire the 'variable scope' deprecated graph mode tests to make them pass. Specifically defining a new variable scope within a function call exercises the kwargs in graph mode decorator which raises an exception. My other option would be to modify the graph mode decorator which I feel might be riskier. I hope that is ok. But if this is a valid user scenario then I guess I need to take that route.  Now all the tests pass on my setup. Hopefully the CI does not grumble!

allenlavoie commented:
For now can we work around the `variables` keyword issue by adding a function in between that converts it to a positional argument?

```
def forward(*args, **kwargs):
  ...
  def grad_wrapper(*output_grads, variables=None):
    @tf.custom_gradient
    def grad(variables, *output_grads):
      ...
    return grad(variables, *output_grads)
```

I agree that fixing the graph-mode decorator is a more satisfying solution. But it'd be good to avoid regressing on recompute_grad if we can work around it.

pidajay commented:
Sounds reasonable. Will work on it.

Commit:
Added grad_wrapper to accomodate graph mode

pidajay commented:
Done! But down to one last test - testFnRecomputeWithScopeGradients. Unfortunately it just hangs without any explanation on my system. Will try to investigate later. Meanwhile if the changes look ok it will be nice to see if the same issue occurs on CI. Thanks again!

alextp reviewed (dismissed):
Approving to trigger CI

pidajay commented:
@alextp Thanks for triggering the CI. But I think the concerned tests are marked as no_oss. These are the gradients_test target in tensorflow/python/BUILD. I think you might need to trigger an internal CI somewhere.

pidajay commented:
@allenlavoie Is there anything I need to do? Did the internal CI pass?

allenlavoie commented on tensorflow/python/keras/integration_test/gradient_checkpoint_test.py:
```diff
@@ -0,0 +1,163 @@
+                  memory_limit=1024)
+          ])
+    except RuntimeError as e:
+      print(e)
```
What kind of RuntimeError is being ignored here? And why is this non-fatal?

The internal CI doesn't like `print`; that's easy to ignore, but I'm wondering if there's a more explicit thing we can do instead.

pidajay replied:
You are right. I think the test should either fail or be skipped if a virtual GPU cannot be created. Should I add a return False in which case the test will be skipped?

allenlavoie replied:
If we don't think it'll happen during CI builds, I'm happy with this being a test failure. I don't see why it would: either it'll have a decent-sized GPU or it won't have anything.

pidajay replied:
Ok. I will just remove the try/except block then and update the PR?

allenlavoie replied:
Sounds good to me.

Commit:
Fixed test

allenlavoie commented on tensorflow/python/keras/integration_test/BUILD:
```diff
@@ -2,6 +2,7 @@
 #   Contains Keras integration tests that verify with other TF high level APIs.
 
 load("//tensorflow:tensorflow.bzl", "tf_py_test")
```
nit: load("//tensorflow:tensorflow.bzl", "tf_py_test", "cuda_py_test")

pidajay replied:
Done!

pidajay replied:
Looks like sanity test was not happy with this. Fixed it. Hopefully things go through this time.

allenlavoie commented on tensorflow/python/keras/integration_test/gradient_checkpoint_test.py:
```diff
@@ -0,0 +1,160 @@
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow.keras import layers, optimizers
```
Can you use:

layers = tf.keras.layers
optimizers = tf.keras.optimizers

it's a tensorflow.python.keras vs. tensorflow.keras issue otherwise. If we always go through the public API I think it'll be fine.

pidajay replied:
Done. Thanks!

Commit:
Addressed PR comments

Commit:
Fixed sanity test issue

pidajay commented:
@allenlavoie Copybara looks unhappy. Is there anything I need to do?

allenlavoie commented:
Ah, yeah, everything else seems OK now but python:gradients_test is timing out on testFnRecomputeWithScopeGradients. It looks like ` self.evaluate(test_var.assign(...` just needs to move outside the TestFn. You can just run tf.global_variables_initializer (looking the variable up in the global collection) or something like that if it's easier.

The issue is that tf.gradients grabs a lock for the graph (since it mutates it), and Session.run grabs the same lock (though self.evaluate, since it requires the graph to stay the same while it Session.runs). The lock isn't reentrant so it deadlocks. The easy fix is just to not run Session.run calls while tf.gradients is mutating the graph.

Commit:
Fixed test

pidajay commented:
Ah, got it. I have pushed the changes. It  seems to work now locally. I will keep my fingers crossed :)

allenlavoie commented on tensorflow/python/ops/gradients_test.py:
```diff
@@ -1442,6 +1481,8 @@ def TestFn(input_t):
       out_re = test_fn_re(test_input_t)
       out = TestFn(test_input_t)
 
+    init = tf.compat.v1.global_variables_initializer()
```
there's no `tf`; variables.global_variables_initializer() should work

pidajay replied:
Very sorry about that. Fixed it. Thanks

Commit:
Fixed test

pidajay commented:
@allenlavoie You might need to jumpstart copybara

pidajay commented:
Do I need to do something on my end regarding the MacOS python2 failure? The errors do not seem related to my changes.

allenlavoie commented:
No action needed; if there are a bunch of unrelated failures it probably means the presubmit has been marked as non-blocking for now. It's running some final tests now, I'll let you know if there are failures.

allenlavoie commented:
In tensorflow/python:gradients_test, VariablesGradientTest.testCustomGradientRecomputeGradHigherOrder says "TypeError: If using @custom_gradient with a function that uses variables, the enclosing variable scope must have use_resource=True."

I think the issue is that it's trying to run with v2 behaviors disabled (among the many other configurations). Can you add a run_v2_only decorator to the test? Like this: https://github.com/tensorflow/tensorflow/blob/3e95e9f03aea5592eef2bcfa6eb8ec94cbb73db3/tensorflow/python/ops/stateful_random_ops_test.py#L92

There's no reason to worry about v1 here, it's just a quirk of the test infrastructure.

Commit:
Fixed test

rafafael03 commented:
How can I update my tensorflow so that I use this new changes?  
And how is the correct way to use it?

What a awesome feature!!

pidajay commented:
Thanks guys for helping get this PR through. Really appreciate all your help and your patience!

pidajay commented:
@rafafael03 Since the PR is merged you should be able to get it from the tf-nightly build (pip install tf-nightly-gpu) from tomorrow I assume. You can find an example of how to use this feature over [here](https://github.com/pidajay/tf2_gradient_checkpointing/blob/master/tf_recompute_grad_tutorial.ipynb). Unfortunately you still have to manually partition/split the model to see the benefits. I had something in the works where this thing can be done automatically without manual split and better memory savings. You can find it [here](https://github.com/pidajay/tf2_gradient_checkpointing/blob/master/gradient_checkpointing.py) with an example how to use it [here](https://github.com/pidajay/addons/blob/grad_checkpointing_eager/docs/tutorials/training_gradient_checkpointing.ipynb). But currently it work only in eager mode for sequential models only. Hopefully if I find some time I can polish it and try to get it over to TF core.

left-brain commented:
Is this feature implemented in tensorflow 2.4.0?

nyngwang commented:
> [..] You can find it [here](https://github.com/pidajay/tf2_gradient_checkpointing/blob/master/gradient_checkpointing.py) with an example how to use it [here](https://github.com/pidajay/addons/blob/grad_checkpointing_eager/docs/tutorials/training_gradient_checkpointing.ipynb). But currently it works only in eager mode for sequential models only. Hopefully if I find some time I can polish it and try to get it over to TF core.

@pidajay Really appreciate your work! It was so amazing when I realized that your repo is merged into TF core! I would like to ask you just a simple little question: Does the tutorial/docs in the last link of your previous comment still work on tensorflow v2.8?
